YunoHost/yunohost-admin
# Enh configpanels

Sorry, another change for config panels to get routing on tabs. This would allow us to add links to a specific tab of a config panel or other panel like pages, for example to point someone to the `cert-management`'s tab of a domain in my upcoming rework of domains page. This also open the door for more control and possible overwrite of panels, per panel (also used in the upcoming rework) and lazy loading of tabs content.  
Also add improvements on routing handling and routing related watchers.

## Changelog
- reworked `<tab-form>` into `<abstract-form>` which should be more generic for use in cards and tabs
- add `<routable-tabs>` component to insert another level of routing inside a view thru a navigable equivalent of `<b-tabs>`.
- reduced part of `<config-panels>` into `<config-panel>` component that can be used as a [child view](https://github.com/YunoHost/yunohost-admin/blob/fa63687cd6b7b8c5bafa9c9bcdb9e304604d22ad/app/src/router/routes.js#L162) on a config panel view (`<config-panel>` is rendered on the nested `<router-link>`)
- add better handling of the main `<router-link>`'s `key` prop which was specified by path, this was actually forcing component replacement every time, it now computes a unique key only for routes with dynamic `params` and allow definition of which params should trigger component replacement (thru `meta.routerParams` key in [route definition](https://github.com/YunoHost/yunohost-admin/blob/fa63687cd6b7b8c5bafa9c9bcdb9e304604d22ad/app/src/router/routes.js#L165)).
- update `breadcrumb` handling and move it to the store (the breadcrumb doesn't show the tab name)


## How to declare a config panel route

see `vue-router` [nested views doc](https://router.vuejs.org/guide/essentials/nested-routes.html) for the syntax.

```js
{
  // no need for name here, only children are visited
  path: '/apps/:id/config-panel',
  component: () => import(/* webpackChunkName: "views/apps/config" */ '@/views/app/AppConfigPanel'),
  props: true,
  // Define children routes (tab/panel)
  children: [
    {
      name: 'app-config-panel',
      // `tabId` will be the current `panel.id` that will be used to display the specified form
      path: ':tabId?',
      // Use the ConfigPanel component as the nested child view
      component: () => import(/* webpackChunkName: "components/configPanel" */ '@/components/ConfigPanel'),
      props: true,
      meta: {
        // Overwrite params used to define the <router-link>'s key (declaring only `id` without `tabId` ensure that current upper view (AppConfigPanel) will not rerender)
        routerParams: ['id'],
        args: { trad: 'app_config_panel' },
        breadcrumb: ['app-list', 'app-info', 'app-config-panel']
      }
    }
  ]
}
```

## HOW TO TEST

go to a domain or an app config page and you should see an url like `/domains/ynh.local/config/feature` and `/domains/ynh.local/config/dns`, reloading on a tab should still display the last tab you were on.

Since the upper view is not recreated at tab change (thanks to `routerParams: ['id']` in routes definition) and that forms data is handle into it, tabs will save form modification even if the tab change.
